### PR TITLE
Feature reshape function

### PR DESCRIPTION
**Pull request by mtar**
## Description

Include a summary of the change/s.
Please also include relevant motivation and context. List any dependencies that are required for this change.

This PR proposes a rudimentary reshape functionality on 1-dim flat tensors returned by some functions (i.e, ht.arange, ht.linspace, and ht.logspace). Additionally, reshaping of all non-distributed tensors is handled by the pytorch function.
~Functionality for other kind of distributed tensors is not implemented~
**EDIT**: It should now work for all tensors with split == 0 now
**EDIT**: All split axes are supported

Issue/s resolved: #143

## Changes proposed:
- local torch.reshape on non-distributed tensors
- reshape on vectors or flat tensors

## Type of change

Remove irrelevant options:
- New feature (non-breaking change which adds functionality)
- Documentation update

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by mtar** [comment]
Support for all tensors with split == 0 added

**Comment by coquelin77** [comment]
have you looked into how to get this to work for `split=1`?

File: heat/core/manipulations.py [hunk 1]
```diff
@@ -699,6 +700,113 @@ def hstack(tup):
...
+        return factories.array(
+            torch.reshape(a._DNDarray__array, shape), dtype=a.dtype, device=a.device, comm=a.comm
+        )
+
+    # Check for currently not supported tensors
+    if a.split > 0:
+        raise NotImplementedError("Tensors with split axes > 0 are not supported yet")
+
+    # Create new flat result tensor
+    _, local_shape, _ = a.comm.chunk(shape, a.split)
```
Comment (coquelin77) [inline heat/core/manipulations.py:785]: can use `a.lshape` instead

File: heat/core/manipulations.py [hunk 2]
```diff
@@ -699,6 +700,113 @@ def hstack(tup):
...
+    data = torch.empty(np.prod(local_shape), dtype=a.dtype.torch_type())
```
Comment (coquelin77) [inline heat/core/manipulations.py:786]: device call, and numpy doesnt run on GPU

File: heat/core/tests/test_dndarray.py [hunk 3]
```diff
@@ -402,6 +402,23 @@ def test_redistribute(self):
             with self.assertRaises(ValueError):
                 st.redistribute_(target_map=torch.zeros((2, 4)))
 
+    def test_reshape(self):
```
Comment (coquelin77) [inline heat/core/tests/test_dndarray.py:405]: these tests can be deleted. you can do the inline call in the manipulations tests. no need to replicate the tests

File: heat/core/tests/test_manipulations.py [hunk 4]
```diff
@@ -874,6 +874,49 @@ def test_hstack(self):
...
+        self.assertTrue(ht.equal(reshaped, result))
+
+        # 1-dim distributed vector
+        a = ht.arange(8, dtype=ht.float32, split=0)
+        result = ht.array([[[0, 1], [2, 3]], [[4, 5], [6, 7]]], dtype=ht.float32, split=0)
+        reshaped = ht.reshape(a, (2, 2, 2))
+
+        self.assertEqual(reshaped.size, result.size)
+        self.assertEqual(reshaped.shape, result.shape)
+        self.assertTrue(ht.equal(reshaped, result))
```
Comment (coquelin77) [inline heat/core/tests/test_manipulations.py:894]: test for dtype

Comment (coquelin77) [reply heat/core/tests/test_manipulations.py:894]: also we use floats as the default so this doesnt test anything different. please change to float64

File: heat/core/manipulations.py (hunk 1 above)
Comment (mtar) [reply heat/core/manipulations.py:785]: I need the new local shape here.

Comment (coquelin77) [reply heat/core/manipulations.py:785]: sorry i missed that. this is correct

File: heat/core/manipulations.py (hunk 2 above)
Comment (mtar) [reply heat/core/manipulations.py:786]: Thanks for the hint. I hadn't the opportunity to test it on the GPU.
I don't really get the second remark. np.prod will return an integer?

**Comment by mtar** [comment]
> have you looked into how to get this to work for `split=1`?

I couldn't come up with a nice solution I'm happy with. So I thought it's better to have some reshape functionality than none  :grinning:

File: heat/core/manipulations.py (hunk 2 above)
Comment (mtar) [reply heat/core/manipulations.py:786]: It runs on GPU. DNDarray handles wrong devices, but, nevertheless, it is better to change it beforehand.

File: heat/core/tests/test_dndarray.py (hunk 3 above)
Comment (mtar) [reply heat/core/tests/test_dndarray.py:405]: Good to know. Thank you.

File: heat/core/manipulations.py (hunk 2 above)
Comment (coquelin77) [reply heat/core/manipulations.py]: the numpy command will run on the CPU process even if the device is GPU, which is why we need to call the torch version. it isnt going to save a lot of time but out programming model is to put it all on one device.

so it would be:

```python
_, local_shape, _ = a.comm.chunk(shape, a.split)
lshp = torch.prod(torch.tensor(local_shape, device=a.device.torch_device), dim=0)
data = torch.empty(local_shape, dtype=a.dtype.torch_type(), device=a.device.torch_device)
```
im not sure if torch requires the dim argument for prod

Comment (mtar) [reply heat/core/manipulations.py]: If you say so, I will change it :confused: 
Does it really save time? It seems more like an overhead for me creating a tensor and pushing data to the gpu and back for the calculation.

Comment (coquelin77) [reply heat/core/manipulations.py]: with the use of torch it does not leave the gpu. when you use numpy and the device is gpu then it has to leave the gpu, go to the cpu, then go back to the gpu after it's done.

**Comment by mtar** [comment]
All split axes are supported now.

File: heat/core/manipulations.py [hunk 5]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+    """
+
+    def reshape_counts_displs(csum1, csum2):
+        """
+        Calculate the counts and displacements needed for redistributing the data from two cumulative sums.
+        """
+        p = csum1[a.comm.rank - 1] if a.comm.rank > 0 else 0
+        i = 0
+
+        result = np.zeros((a.comm.size,), dtype=np.int)
```
Comment (coquelin77) [inline heat/core/manipulations.py:784]: torch not numpy

File: heat/core/manipulations.py [hunk 6]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+                maxpos = i
+
+        # write counts
+        while csum2[i] < csum1[a.comm.rank]:
+            result[i] = csum2[i] - p
+            p = csum2[i]
+            i += 1
+        result[i] = csum1[a.comm.rank] - p
+
+        return tuple(result), (0,) + tuple(np.cumsum(result[:-1]))
```
Comment (coquelin77) [inline heat/core/manipulations.py:807]: torch not numpy

File: heat/core/manipulations.py [hunk 7]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+
+    def reshape_argsort_counts_displs(shape1, lshape1, displs1, shape2, displs2, axis, comm):
+        """
+        Compute the send order, counts, and displacements.
+        """
+        mask = torch.zeros(lshape1).flatten()
```
Comment (coquelin77) [inline heat/core/manipulations.py:813]: dtype, device, or use `zeros_like`

File: heat/core/manipulations.py [hunk 8]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+
+        local_index = 0
+        global_index = displs1[comm.rank] * np.prod(shape1[axis + 1 :])
+        local_len = np.prod(lshape1[axis:])
+        global_len = np.prod(shape1[axis:])
+
+        ulen = np.prod(shape2[axis + 1 :])
+
+        while local_index < np.prod(lshape1):
```
Comment (coquelin77) [inline heat/core/manipulations.py:822]: torch not numpy

File: heat/core/manipulations.py [hunk 9]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+                # search new process id
+                for j, k in zip(displs2, range(comm.size)):
+                    if j <= jpos:
+                        mask[local_index + i] = k
+            global_index += global_len
+            local_index += local_len
+
+        uniques, counts = torch.unique(mask, return_counts=True)
+
+        mpicounts = torch.zeros(comm.size)
```
Comment (coquelin77) [inline heat/core/manipulations.py:835]: device + dtype

File: heat/core/manipulations.py [hunk 10]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+
+        for i, j in zip(uniques, counts):
+            mpicounts[int(i)] = int(j)
+
+        return (
+            mask.argsort(),
+            tuple(mpicounts.numpy()),
```
Comment (coquelin77) [inline heat/core/manipulations.py:842]: torch not numpy

File: heat/core/manipulations.py [hunk 11]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+            (0,) + tuple(torch.cumsum(mpicounts[:-1], dim=0).numpy()),
+        )
+
+    # Check the type of shape and number elements
+    shape = stride_tricks.sanitize_shape(shape)
+    if np.prod(shape) != a.size:
```
Comment (coquelin77) [inline heat/core/manipulations.py:848]: torch not numpy

File: heat/core/manipulations.py [hunk 12]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+            mpicounts[int(i)] = int(j)
+
+        return (
+            mask.argsort(),
+            tuple(mpicounts.numpy()),
+            (0,) + tuple(torch.cumsum(mpicounts[:-1], dim=0).numpy()),
+        )
+
+    # Check the type of shape and number elements
+    shape = stride_tricks.sanitize_shape(shape)
```
Comment (coquelin77) [inline heat/core/manipulations.py:847]: shape is ambiguous in this specific function. which shape do you mean here?

File: heat/core/manipulations.py [hunk 13]
```diff
@@ -735,6 +736,170 @@ def hstack(tup):
...
+        a.comm.Alltoallv((send.flatten(), sendcounts, senddispls), (data, recvcounts, recvdispls))
+
+        # original order
+        backsort = torch.argsort(recvsort)
+        data = data[backsort]
+
+    else:
+        # turn displs into cumulative sums
+        old_csum = tuple(np.prod(a.shape[1:]) * i for i in old_displs[1:]) + (a.size,)
+        new_csum = tuple(np.prod(shape[1:]) * i for i in new_displs[1:]) + (a.size,)
```
Comment (coquelin77) [inline heat/core/manipulations.py:886]: torch not numpy

**Comment by mtar** [comment]
I have problems with some matrices. I'm getting a wrong order.

**Comment by mtar** [comment]
argsort is causing this. On more than 11 elements the order is lost.

File: heat/core/manipulations.py (hunk 13 above)
Comment (mtar) [reply heat/core/manipulations.py]: There are now a bunch of torch function calls

**Comment by mtar** [comment]
Axis parameter added. It is now possible to set a different split axis for the new shape.
